src/context/fetch.ts

```diff
@@ -26,14 +28,24 @@ const createFetchRequestParameters = (input: MockedRequest): RequestInit => {
     return requestParameters
   }
 
-  if (
-    typeof body === 'object' ||
-    typeof body === 'number' ||
-    typeof body === 'boolean'
-  ) {
-    requestParameters.body = JSON.stringify(body)
-  } else {
-    requestParameters.body = body
+  switch (typeof body) {
+    case 'object':
+      if (input.headers?.get('content-type')?.includes('multipart/form-data')) {
+        const formData = new FormData()
```

Comment:
There are two issues with our approach here. 

1. It will only work in the browser. `FormData` doesn't exit in Node.js, so this line will throw. 
2. I think we're trying to handle a scenario much later than we should be doing that. 

It looks like we should re-create a `FormData` when parsing intercepted request body right here:

https://github.com/mswjs/msw/blob/6fc89f762681aa7f6f697ea78fa5f7c736b35c5d/src/utils/request/parseBody.ts#L16-L21

I think the fix should be somewhere in `parseBody` or `parseMultipartData`. 

I may be wrong on this though. Do you find this request body processing chain to be correct?

1. Request is intercepted, its body is text because we only transfer text via WebSocket channel or Node.js interceptor.
2. Request body is parsed in `parseBody`. We see it's multipart form data, we convert the string to a JSON representation of the form data.
3. We pass the converted JSON request body to `req.body`, it arrives to `ctx.fetch`.
4. We reconstruct `FormData` from the JSON request body and dispatch the native fetch.

It looks like if we reconstruct `FormData` in `parseBody` then we can drop any extra body processing in `ctx.fetch`. This only comes down to one question:

- What do you expect `req.body` to be when intercepted a `multipart/form-data` request? Do you expect it to be `FormData`? Or a JSON representation of the form data?

Reply:
> What do you expect req.body to be when intercepted a multipart/form-data request? Do you expect it to be FormData? Or a JSON representation of the form data?

Do you mean inside a user defined `RequestHandler`?  That's an interesting question.

Without `msw`:
1. User calls native `fetch`: ` fetch('/post', { method: 'POST', body: formData })`
2. The browser modifies the request e.g. adds `Content-Type` header, changes body to a string: `"------WebKitFormBoundaryxOsEsNP9Pphx5RC6\r\nContent-Disposition: form-data; name=\"key\"\r\n\r\nvalue\r\n------WebKitFormBoundaryxOsEsNP9Pphx5RC6--\r\n"` etc.
3. Browser sends request to the backend

With `msw`:

4. The service worker intercepts the request in step 3 and passes it in a 4th step to the user's `RequestHandler`.

If one is not an expert in service workers or `msw`, it would be reasonable to expect `RequestHandler` to receive either
- the original request from step 1 without any modifications (has no headers, has `FormData` body)
- whatever is send to the backend (result of step 2)

However what is actually happening is that `msw` modifies the request again between step 3 and 4 (to make usage in `RequestHandler` more convenient?)  e.g. by changing `.headers` from a plain object to `new Headers` in `parseWorkerRequest.ts`.

So I guess it would be valid to change the body back to `FormData` for convenience as well. 

On the other hand `msw` tries to keep `MockedRequest` in the browser and NodeJS equivalent (Why does `parseWorkerRequest` return `RestRequest` with an extra `params` property?), so users can work with requests in the same way, which speaks for a plain object body in both environments.

But then again how useful is that equivalence in the case of NodeJS anyways? Without `FormData` you cannot write your unit tests in the same way the browser makes the fetch request anyways.

Maybe going for changing the body back to `FormData` is more in line with the current approach? What about also adding a polyfill for FormData?

Very curious about what you expect (JSON or FormData)!

Reply:
Probably a bad idea with a lot of breaking changes, but would it be possible to wrap the native `fetch` (as it is done while waiting for the serviceworker to initialise in `deferNetworkRequestsUntil.ts` ) in a function `keepOriginalThenFetch`, which stores the original request from step 1 and then calls the native fetch. Then in `parseWorkerRequest` we get the request modified by the browser from the service worker and create the `RestRequest` the same way as before (keeping the modifications made by the browser in step 2), but afterwards we retrieve the original request from step 1 stored by `keepOriginalThenFetch` and overwrite the properties of `RestRequest`. Then `RestRequest` would have the original body from step 1 and all other original user specified properties such as `headers`.

https://github.com/mswjs/msw/issues/1158 was about `FormData` only, but what about all the other types supported by `BodyInit`, which are missing in`DefaultBodyType`?
```ts
type BodyInit = Blob | BufferSource | FormData | URLSearchParams | ReadableStream<Uint8Array> | string;
```
With `keepOriginalThenFetch` all types would work and there would not be any need to parse the body or call `parseMultipartData` any more.

I haven't used `msw` enough or even made a POST request with `BufferSource` in my life, so I don't really know what I am talking about here 😅